src/main/java/delight/nashornsandbox/internal/GraalSandboxImpl.java

```diff
@@ -0,0 +1,127 @@
+		super(GraalJSScriptEngine.create(null, Context.newBuilder().allowExperimentalOptions(true).allowPolyglotAccess(PolyglotAccess.ALL).allowHostAccess(HostAccess.ALL)), params);
+		isStrict = Arrays.asList(params).contains("-strict");
+		Bindings bindings = this.scriptEngine.getBindings(ScriptContext.ENGINE_SCOPE);
+		bindings.put("polyglot.js.allowHostClassLookup", (Predicate<String>) s -> this.sandboxClassFilter.getStringCache().contains(s));
+	}
+	
+	@Override
+	protected void resetEngineBindings() {
+		
+    }
+	
+	@Override
+	public Bindings createBindings() {
+		return scriptEngine.getBindings(ScriptContext.ENGINE_SCOPE);
+	}
+	
+    protected void produceSecureBindings() {
+        try {
+            final StringBuilder sb = new StringBuilder();
+            cached = scriptEngine.getBindings(ScriptContext.ENGINE_SCOPE);
+            sanitizeBindings(cached);
+            if (!allowExitFunctions) {
+                sb.append("var quit=function(){};var exit=function(){};");
+            }
+            if (!allowPrintFunctions) {
+                sb.append("var print=function(){};var echo = function(){};");
+            }
+            if (!allowReadFunctions) {
+                sb.append("var readFully=function(){};").append("var readLine=function(){};");
+            }
+            if (!allowLoadFunctions) {
+                sb.append("var load=function(){};var loadWithNewGlobal=function(){};");
+            }
+            if (!allowGlobalsObjects) {
+                // Max 22nd of Feb 2018: I don't think these are strictly necessary since they are only available in scripting mode
+                sb.append("var $ARG=null;var $ENV=null;var $EXEC=null;");
+                sb.append("var $OPTIONS=null;var $OUT=null;var $ERR=null;var $EXIT=null;");
+            }
+            scriptEngine.eval(sb.toString());
+            this.engineAsserted = true;
+        }
+        catch (final Exception e) {
+            throw new RuntimeException(e);
+        }
+    }
+    
+    @Override
+	protected Bindings secureBindings(Bindings bindings) {
+        if (bindings == null)
+            return null;
+
+        if (bindings != cached) {
+        	for (Map.Entry<String, Object> entry : bindings.entrySet()) {
+        	  cached.putIfAbsent(entry.getKey(), entry.getValue());
+        	}
+        } else { 
+        	cached.putAll(bindings);
+        }
+        return cached;
+    }
+    
+	@Override
+	public Object eval(final String js, final ScriptContext scriptContext, final Bindings bindings)
+			throws ScriptCPUAbuseException, ScriptException {
+		if (scriptContext != null) {
+			Bindings engineBindings = scriptEngine.getBindings(ScriptContext.ENGINE_SCOPE);
+			Bindings contextBindings = scriptContext.getBindings(ScriptContext.ENGINE_SCOPE);
+			if (contextBindings != null) {
+				for (String key : contextBindings.keySet()) {
+					if (engineBindings.get(key) != null) contextBindings.remove(key);
+				}			
+			}
+		}
```

Comment:
What do these lines aim to do, and if they meet a need why not have it in the other method?

Reply:
If a script context is provided, its bindings will be evaluated later - i.e. inside the script engine - and merged into the engine bindings. Nashorn automatically checks against global bindings but Graal seems to override global binding entries if the same key is present in the script context's bindings. This will prevent overriding globals. We don't know if this is intentional behavior by GraalJS.